File: code/872/solution.go
package main

import "fmt"

func f(a, b int64) int64 {
	if a <= b {
		return a
	}
	le, ri := b, b
	for i := int64(1); ; i *= 2 {
		le = ri + 1
		ri = ri + i
		if ri >= a {
			return b + f(a, le)
		}
	}
}

func xpow(a, b int64) int64 {
	ret := int64(1)
	for i := int64(0); i < b; i++ {
		ret *= a
	}
	return ret
}

func main() {
	fmt.Println(f(6, 1))
	fmt.Println(f(10, 3))
	fmt.Println(f(xpow(10, 17), xpow(9, 17)))
}

File: code/872/solution.cpp
#include <iostream>
using namespace std;
typedef long long ll;
ll f(ll a, ll b) {
    if (a <= b)
        return a;
    ll le = b, ri = b;
    for (ll i = 1;; i *= 2) {
        le = ri + 1;
        ri = ri + i;
        if (ri >= a) {
            return b + f(a, le);
        }
    }
}

inline ll xpow(ll a, ll b) {
    ll ret = 1;
    for (int i = 0; i < b; i++)
        ret *= a;
    return ret;
}

int main() {
    cout << f(6, 1) << endl;
    cout << f(10, 3) << endl;
    cout << f(xpow(10, 17), xpow(9, 17)) << endl;
    return 0;
}
